File: mobile/ios/packages/town-crier-domain/Sources/ValueObjects/WatchZoneLimits.swift
/// Encodes the watch zone limits for a subscription tier.
///
/// Zone count limits delegate to ``EntitlementMap`` so there is a single source of truth
/// that stays in sync with the API's `EntitlementMap.cs`.
public struct WatchZoneLimits: Equatable, Sendable {
  public let tier: SubscriptionTier
  public let maxZones: Int
  public let maxRadiusMetres: Double
  /// Whether this tier can draw a custom-shape (polygon) zone rather than
  /// only a circle (GH#1031). Mirrors the server's
  /// `SubscriptionTier.AllowsCustomBoundary()` (`t.IsPaid()`).
  public let allowsCustomBoundary: Bool

  public init(tier: SubscriptionTier) {
    self.tier = tier
    maxZones = EntitlementMap.limit(for: tier, quota: .watchZones)
    switch tier {
    case .free:
      maxRadiusMetres = 2000
      allowsCustomBoundary = false
    case .personal:
      maxRadiusMetres = 5000
      allowsCustomBoundary = true
    case .pro:
      maxRadiusMetres = 10000
      allowsCustomBoundary = true
    }
  }

  /// Whether the user can add another zone given their current count.
  public func canAddZone(currentCount: Int) -> Bool {
    currentCount < maxZones
  }

  /// Clamps a radius to the tier's maximum.
  public func clampRadius(_ radius: Double) -> Double {
    min(radius, maxRadiusMetres)
  }
}

File: mobile/android/domain/src/main/kotlin/uk/towncrierapp/domain/watchzones/WatchZoneLimits.kt
package uk.towncrierapp.domain.watchzones

import uk.towncrierapp.domain.subscriptions.EntitlementMap
import uk.towncrierapp.domain.subscriptions.Quota
import uk.towncrierapp.domain.subscriptions.SubscriptionTier

/**
 * Per-tier watch-zone limits. Zone count delegates to [EntitlementMap] so
 * there is a single source of truth kept in sync with the API's
 * `EntitlementMap.cs`; the radius cap is this type's own table and is
 * client-enforced only — the server ceiling is a flat 10 km for every tier
 * (bump the server first before ever raising these). Port of iOS
 * `WatchZoneLimits`.
 */
public data class WatchZoneLimits(
    public val tier: SubscriptionTier,
) {
    public val maxZones: Int = EntitlementMap.limit(tier, Quota.WATCH_ZONES)

    public val maxRadiusMetres: Double =
        when (tier) {
            SubscriptionTier.FREE -> 2_000.0
            SubscriptionTier.PERSONAL -> 5_000.0
            SubscriptionTier.PRO -> 10_000.0
        }

    /** Whether the user can add another zone given their current count. */
    public fun canAddZone(currentCount: Int): Boolean = currentCount < maxZones

    /** Clamps a radius to the tier's maximum. */
    public fun clampRadius(radiusMetres: Double): Double = minOf(radiusMetres, maxRadiusMetres)
}
